demos/sslecho/main.c

```diff
@@ -0,0 +1,356 @@
+            {
+                perror("Unable to accept");
+                exit(EXIT_FAILURE);
+            }
+
+            printf("Client TCP connection accepted\n");
+
+            // Create server SSL structure using the new socket.
+            ssl = SSL_new(ssl_ctx);
+            SSL_set_fd(ssl , client_skt);
+
+            // Wait for SSL connection from the client.
+            if (SSL_accept(ssl) <= 0)
+            {
+                ERR_print_errors_fp(stderr);
+            }
+            else
+            {
+
+                printf("Client SSL connection accepted\n\n");
+
+                // Echo loop.
+                while (true)
+                {
+                    // Get message from client; will fail if client closes connection.
+                    if ((rxlen = SSL_read(ssl , rxbuf , rxcap)) <= 0)
+                    {
+                        if (rxlen == 0)
+                        {
+                            printf("Client closed connection\n");
+                        }
+                        ERR_print_errors_fp(stderr);
+                        break;
+                    }
+                    // Insure null terminated input.
+                    rxbuf[rxlen] = 0;
+                    // Look for kill switch.
+                    if (strcmp(rxbuf , "kill\n") == 0)
+                    {
+                        // Terminate...with extreme prejudice.
+                        printf("Server received 'kill' command\n");
+                        server_running = false;
+                        break;
+                    }
+                    // show received message.
+                    printf("Received: %s" , rxbuf);
+                    // Echo it back
+                    if (SSL_write(ssl , rxbuf , rxlen) <= 0)
+                    {
+                        ERR_print_errors_fp(stderr);
+                    }
+                }
+            }
+            if (server_running)
+            {
+                // Cleanup for next client.
+                SSL_shutdown(ssl);
+                SSL_free(ssl);
+                close(client_skt);
+            }
+        }
+        printf("Server exiting...\n");
+    }
+    // Else client.
+    else
+    {
+        printf("We are the client\n\n");
+
+        // Create "bare" socket.
+        client_skt = create_socket(false);
+        // Set up connect address.
+        addr.sin_family = AF_INET;
+        inet_pton(AF_INET , rem_server_ip , &addr.sin_addr.s_addr);
+        addr.sin_port = htons(server_port);
+        // Do TCP connect with server.
+        if (connect(client_skt , (struct sockaddr*)&addr , sizeof(addr)) != 0)
+        {
+            perror("Unable to TCP connect to server");
+            goto Exit;
+        }
+        else
+        {
+            printf("TCP connection to server successful\n");
+        }
+
+        // Create server SSL structure using client socket.
+        ssl = SSL_new(ssl_ctx);
+        SSL_set_fd(ssl , client_skt);
```

Comment:
Best practice is to set SNI here using `SSL_set_tlsext_host_name`.

Reply:
Not understanding this...
From the OpenSSL docs: int SSL_set_tlsext_host_name(const SSL *s, const char *name);
Does this mean that the above function replaces the TCP accept() that returned client_skt?
This code came from the Wiki example. And it does work...

Reply:
No, this doesn't replace anything in the existing code. Just call the `SSL_set_tlsext_host_name` function *additionally* after the `SSL_set_fd` call. This causes the hostname of the target server to be sent in the TLS ClientHello so that a server which hosts multiple domains behind a single IP can return the correct certificate. This is not *required* by the standard, but is considered best practice. Many servers these days will actually fail to offer a suitable connection without this data.

Reply:
Thanks.